presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PartialAggregationPushDown.java

```diff
@@ -79,17 +77,114 @@ public Rewriter(SymbolAllocator allocator, PlanNodeIdAllocator idAllocator)
-            if (!decomposable || !allowPushThrough(node.getSource())) {
+
+            if (!decomposable) {
                 return context.defaultRewrite(node);
             }
 
+            if (node.getStep() == SINGLE) {
+                return context.rewrite(split(node));
+            }
+            else if (node.getStep() == PARTIAL) {
+                return context.rewrite(pushPartial(node, exchange));
+            }
+
+            return context.defaultRewrite(node);
+        }
+
+        private PlanNode pushPartial(AggregationNode aggregation, ExchangeNode exchange)
+        {
+            ImmutableList.Builder<List<Symbol>> exchangeAssignments = ImmutableList.builder();
+            ImmutableList.Builder<PlanNode> partials = ImmutableList.builder();
+            for (int i = 0; i < exchange.getSources().size(); i++) {
+                PlanNode source = exchange.getSources().get(i);
+
+                if (!exchange.getOutputSymbols().equals(exchange.getInputs().get(i))) {
+                    // add an identity projection to preserve the inputs to the aggregation, if necessary
+                    ImmutableMap.Builder<Symbol, Expression> assignments = ImmutableMap.builder();
+                    for (int outputIndex = 0; outputIndex < exchange.getOutputSymbols().size(); outputIndex++) {
+                        Symbol output = exchange.getOutputSymbols().get(outputIndex);
+                        assignments.put(output, exchange.getInputs().get(i).get(outputIndex).toSymbolReference());
+                    }
+
+                    source = new ProjectNode(idAllocator.getNextId(), source, assignments.build());
+                }
+
+                partials.add(new AggregationNode(
+                        idAllocator.getNextId(),
+                        source,
+                        aggregation.getAggregations(),
+                        aggregation.getFunctions(),
+                        aggregation.getMasks(),
+                        aggregation.getGroupingSets(),
+                        aggregation.getStep(),
+                        aggregation.getHashSymbol(),
+                        aggregation.getGroupIdSymbol()));
+
+                exchangeAssignments.add(aggregation.getOutputSymbols());
+            }
+
+            PartitioningScheme partitioning = new PartitioningScheme(
+                    exchange.getPartitioningScheme().getPartitioning(),
+                    aggregation.getOutputSymbols(),
+                    exchange.getPartitioningScheme().getHashColumn(),
+                    exchange.getPartitioningScheme().isReplicateNulls(),
+                    exchange.getPartitioningScheme().getBucketToPartition());
+
+            return new ExchangeNode(
+                    idAllocator.getNextId(),
+                    exchange.getType(),
+                    exchange.getScope(),
+                    partitioning,
+                    partials.build(),
+                    exchangeAssignments.build());
```

Comment:
With the validation above, you can change this line to:
```java
                    ImmutableList.copyOf(Collections.nCopies(partials.build().size(), exchange.getOutputSymbols())));
```
and then remove the `exchangeAssignments` variable